### PR TITLE
Image builder GitHub workflow integration design proposal.

### PR DESCRIPTION
**Pull request by dekiel**
**Description**

Changes proposed in this pull request:

Concept proposal for integration of GitHub workflow and ADO oci-image-builder backend. The document describe how image-builder client can be used in GitHub workflow to trigger oci-image-builder pipeline backend to build image. Additionally it describe required changes in scenario when image-builder client is running in prowjob in order to trigger oci-image-builder backend.

### CODE REVIEW

File: cmd/image-builder/github-workflow-integration.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+These parameters include:
+
+### User Defined Parameters
+
+- `Context`: The context of the build.
+- `Dockerfile`: The Dockerfile to be used for the build.
+- `Name`: The name of the image.
+- `BuildArgs`: The build arguments to be passed to the build.
+- `Tags`: The tags to be applied to the image.
+- `ExportTags`: Whether to export the tags.
```
Comment (Sawthis) [inline cmd/image-builder/github-workflow-integration.md:97]: What does it mean to export the tag?

File: cmd/image-builder/github-workflow-integration.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+### Parameters from OIDC Token
+
+- `RepoName`: The name of the repository.
+- `RepoOwner`: The owner of the repository. Possible values include "kyma-project" and "kyma-incubator".
```
Comment (Sawthis) [inline cmd/image-builder/github-workflow-integration.md:102]: kyma-incubator will be owned by compass team and I am not sure if they should be included into new image builder as they should migrate out of Prow until end of June. I think this is question for @TorstenD-SAP .

**Review by Sawthis — CHANGES_REQUESTED** [review]

File: cmd/image-builder/github-workflow-integration.md (hunk 1 above)
Comment (dekiel) [reply cmd/image-builder/github-workflow-integration.md:97]: From image-builder.

```
        flagSet.BoolVar(&o.exportTags, "export-tags", false, "Export parsed tags as build-args into dockerfile. Each tag will have format TAG_x, where x is the tag name passed along with the tag")
```

File: cmd/image-builder/github-workflow-integration.md (hunk 2 above)
Comment (dekiel) [reply cmd/image-builder/github-workflow-integration.md:102]: When `kyma-incubator` org will not be allowed to build images with our solution, we will simply remove respective value from allowed org list.

**Review by Sawthis — APPROVED** [review]

**Review by KacperMalachowski — APPROVED** [review]

File: cmd/image-builder/github-workflow-integration.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+- `kid`: The key ID of the key used to sign the token.
+- `alg`: The algorithm used to sign the token.
+- `run_id`: The ID of the workflow run.
+- `run_number`: The number of the workflow run.
+- `actor`: The login of the user who initiated the workflow run.
+- `event_name`: The name of the event that triggered the workflow run.
+- `workflow`: The name of the workflow that triggered the workflow run.
+- `workflow_ref`: The git ref associated with the workflow file.
+- `repository`: The repository where the workflow run occurred.
+- `repository_owner`: The owner of the repository where the workflow run occurred.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:65]: Wouldn't the present form `occurs` be better in these descriptions?

File: cmd/image-builder/github-workflow-integration.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+### Source Code Cloning Claims
+
+The OIDC token also contains claims that can be used to clone the appropriate version of the source code:
+
+- `repository`: The repository where the workflow run occurred.
+- `repository_owner`: The owner of the repository where the workflow run occurred.
+- `event_name`: The name of the event that triggered the workflow run.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:73]: Please consider using the present forms of the verbs.

File: cmd/image-builder/github-workflow-integration.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+- `ref`: The git ref associated with the workflow run. This can be used to checkout the correct branch, tag, or commit.
+- `ref_type`: The type of git ref associated with the workflow run. This can be used to determine whether the ref is a branch, tag, or
+  commit.
+- `base_ref`: The base git ref associated with the workflow run. This can be used to determine the base branch for a pull request.
+- `head_ref`: The head git ref associated with the workflow run. This can be used to determine the head branch for a pull request.
+
+These claims ensure that the `oci-image-builder` pipeline builds the exact version of the code that was provided in the PR or merged to the
+branch, adhering to slc-29 compliance.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:81]:
```suggestion
branch, adhering to SLC-29 compliance.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+## Parameters for the `oci-image-builder` Pipeline
+
+The `oci-image-builder` pipeline requires certain data to be provided in parameters to execute a build process.
+Certain parameters need to be defined by the user additionally to the data taken from OIDC token.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:86]:
```suggestion
Certain parameters need to be defined by the user in addition to the data taken from the OIDC token.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+These parameters include:
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:88]:
```suggestion
These parameters include user-defined parameters, parameters from the OIDC token, and computed parameters.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+### User Defined Parameters
+
+- `Context`: The context of the build.
+- `Dockerfile`: The Dockerfile to be used for the build.
+- `Name`: The name of the image.
+- `BuildArgs`: The build arguments to be passed to the build.
+- `Tags`: The tags to be applied to the image.
+- `ExportTags`: Whether to export the tags.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:97]:
```suggestion
- **Context**: The context of the build.
- **Dockerfile**: The Dockerfile to be used for the build.
- **Name**: The name of the image.
- **BuildArgs**: The build arguments to be passed to the build.
- **Tags**: The tags to be applied to the image.
- **ExportTags**: Whether to export the tags.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+### Parameters from OIDC Token
+
+- `RepoName`: The name of the repository.
+- `RepoOwner`: The owner of the repository. Possible values include "kyma-project" and "kyma-incubator".
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:102]:
```suggestion
- **RepoName**: The name of the repository.
- **RepoOwner**: The owner of the repository. Possible values include `kyma-project` and `kyma-incubator`.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+- `JobType`: The type of job. Possible values include "presubmit" and "postsubmit".
+- `PullBaseSHA`: The base SHA of the pull request.
+- `PullPullSHA`: The SHA of the pull request.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:105]:
```suggestion
- **JobType**: The type of job. Possible values include `presubmit` and `postsubmit`.
- **PullBaseSHA**: The base SHA of the pull request.
- **PullPullSHA**: The SHA of the pull request.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+### Computed Parameters
+
+The data for this parameter is not available in OIDC token and should not be provided by user.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:109]:
```suggestion
The data for this parameter is not available in the OIDC token and should not be provided by the user.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+We can extract it from `ref` claim of OIDC token.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:110]:
```suggestion
We can extract it from the `ref` claim of the OIDC token.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+- `PullNumber`: The number of the pull request.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:112]:
```suggestion
- **PullNumber**: The number of the pull request.
```

File: cmd/image-builder/github-workflow-integration.md (hunk 3 above)
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:65]: If you can provide values for these claims, please use bold font for them (like you do for [parameters and variables](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#code-font-and-bold-font)).

File: cmd/image-builder/github-workflow-integration.md (hunk 4 above)
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:73]: If you can provide values for these claims, please use bold font for them (like you do for [parameters and variables](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#code-font-and-bold-font)).

File: cmd/image-builder/github-workflow-integration.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+The OIDC token also contains claims that can be used to clone the appropriate version of the source code:
+
+- `repository`: The repository where the workflow run occurred.
+- `repository_owner`: The owner of the repository where the workflow run occurred.
+- `event_name`: The name of the event that triggered the workflow run.
+- `ref`: The git ref associated with the workflow run. This can be used to checkout the correct branch, tag, or commit.
+- `ref_type`: The type of git ref associated with the workflow run. This can be used to determine whether the ref is a branch, tag, or
+  commit.
+- `base_ref`: The base git ref associated with the workflow run. This can be used to determine the base branch for a pull request.
+- `head_ref`: The head git ref associated with the workflow run. This can be used to determine the head branch for a pull request.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:78]: If you can provide values for these claims, please use bold font for them (like you do for [parameters and variables](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#code-font-and-bold-font)).

File: cmd/image-builder/github-workflow-integration.md [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+
+The data for this parameter is not available in OIDC token and should not be provided by user.
+We can extract it from `ref` claim of OIDC token.
+
+- `PullNumber`: The number of the pull request.
+
+## ProwJob Validation in Image Builder
+
+In the scenario where the `oci-image-builder` pipeline is triggered from a ProwJob instead of a GitHub workflow, the validation process
+differs slightly due to the absence of an OIDC token. Triggering the `oci-image-builder` pipeline from a ProwJob is considered a depraecated
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:117]:
```suggestion
differs slightly due to the absence of an OIDC token. Triggering the `oci-image-builder` pipeline from a ProwJob is considered a deprecated
```

File: cmd/image-builder/github-workflow-integration.md [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+scenario and must be supported only for backward compatibility and migration purposes.
+
+### Process Flow
+
+1. **Trigger ProwJob**: User or automation triggers a ProwJob.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:122]:
```suggestion
1. **Trigger ProwJob**: The user or automation triggers a ProwJob.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+2. **Trigger oci-image-builder pipeline**: The ProwJob running Image builder client calls ADO API to trigger `oci-image-builder` pipeline.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:124]:
```suggestion
2. **Trigger oci-image-builder pipeline**: The ProwJob running Image builder client calls ADO API to trigger the `oci-image-builder` pipeline.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+   The parameters required by the `oci-image-builder` pipeline are passed as parameters to the pipeline. These parameters values are
+   collected from data defined by the user and Prow environment variables. Together with the parameters, parameters hash signed by Prow is
+   also passed to the pipeline as a parameter.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:127]:
```suggestion
   The parameters required by the `oci-image-builder` pipeline are passed as parameters to the pipeline. These parameters' values are
   collected from data defined by the user and Prow environment variables. Together with these parameters, the parameters' hash signed by Prow is
   also passed to the pipeline as a parameter.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+
+3. **Validating the Parameters**: The `oci-image-builder` pipeline, running in Azure DevOps, detects the call originates from Prow and
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:129]:
```suggestion
3. **Validating the Parameters**: The `oci-image-builder` pipeline, running in Azure DevOps, detects the call originating from Prow and
```

File: cmd/image-builder/github-workflow-integration.md [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+   validates the parameters against a hash signed by Prow. This step ensures that the parameters are valid and have not been tampered with.
+   The absence of an OIDC token is used to detect that the call is coming from a ProwJob.
+
+### Parameters Validation
+
+The parameters passed to the `oci-image-builder` pipeline are validated against a hash signed by Prow. This hash is computed based on the
+names and values of the parameters. The hash value is signed with secret know to the `Prow` and `oci-image-buidler` pipeline only.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:136]:
```suggestion
The parameters passed to the `oci-image-builder` pipeline are validated against a hash signed by Prow. This hash is computed based on the
names and values of the parameters. The hash value is signed with a secret known to the `Prow` and the `oci-image-buidler` pipeline only.
```

File: cmd/image-builder/github-workflow-integration.md [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+## Activity Diagram
+
+![image-builder-activity-diagram](documentation_assets/image-builder-activity-diagram.png)
+
+## Conclusion
+
+The Image Builder solution, with its seamless integration with GitHub workflows and Azure DevOps pipeline, offers developers a robust and
+secure method to incorporate the building of OCI images into their workflows. By leveraging a signed JWT format in which an OIDC token from
+GitHub's OIDC identity provider is passed, it ensures the secure and authorized passing of information about the workflow and the image to
+build. The entire build process adheres to slc-29 compliance, providing a reliable infrastructure for the building of OCI images.
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:156]:
```suggestion
build. The entire build process adheres to SLC-29 compliance, providing a reliable infrastructure for the building of OCI images.
```

File: cmd/image-builder/github-workflow-integration.md (hunk 3 above)
Comment (dekiel) [reply cmd/image-builder/github-workflow-integration.md:65]: Values for claims are set by GitHub OIDC identity provider.

File: cmd/image-builder/github-workflow-integration.md (hunk 4 above)
Comment (dekiel) [reply cmd/image-builder/github-workflow-integration.md:73]: Values for claims are set by GitHub OIDC identity provider.

**Review by Sawthis — APPROVED** [review]

**Review by KacperMalachowski — DISMISSED** [review]

**Review by KacperMalachowski — DISMISSED** [review]

**Review by Sawthis — DISMISSED** [review]

File: cmd/image-builder/github-workflow-integration.md [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+merged is built.
+
+The validity and integrity of the OIDC token must be validated in the `oci-image-builder` pipeline and fail the pipeline execution if validation fails.
+Because the OIDC token uses the `JWT` format, it can be validated with a standard validation process against GitHub's OIDC identity provider.
+
+### Workflow Identification Claims
+
+The OIDC token contains the following claims that can be used to identify the workflow that triggered the build pipeline. These include:
+
+Sure, here's how you can format the selected text as an unordered list in Markdown, with the words before the colon surrounded by backticks:
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:51]: I guess this isn't part of the doc.

File: cmd/image-builder/github-workflow-integration.md [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+- `workflow`: The name of the workflow that triggers the workflow run.
+- `workflow_ref`: The git ref associated with the workflow file.
+- `repository`: The repository where the workflow run occurs.
+- `repository_owner`: The owner of the repository where the workflow run occurs.
+
+### Source Code Cloning Claims
+
+The OIDC token also contains claims that can be used to clone the appropriate version of the source code:
+
+Sure, here's the selected text rewritten with present tense verbs, while keeping the formatting:
```
Comment (IwonaLanger) [inline cmd/image-builder/github-workflow-integration.md:75]: Same here - the comment doesn't belong here.

File: cmd/image-builder/github-workflow-integration.md (hunk 14 above)
Comment (IwonaLanger) [reply cmd/image-builder/github-workflow-integration.md:84]: What I meant is, that claims are similar to parameters or variables In the sense that they can have some values (no matter where the values come from). Hence, they should be formatted in the same way, i.e. in bold font.

**Review by IwonaLanger — APPROVED** [review]

**Review by KacperMalachowski — APPROVED** [review]

**Pull request merged**